sdk/src/transaction_context.rs

```diff
@@ -826,24 +826,55 @@ impl<'a> BorrowedAccount<'a> {
         Ok(self.account.data_as_mut_slice())
     }
 
-    /// Overwrites the account data and size (transaction wide)
+    /// Overwrites the account data and size (transaction wide).
+    ///
+    /// Call this when you have an owned buffer and want to replace the account
+    /// data with it.
+    ///
+    /// If you have a slice, use set_data_from_slice().
     #[cfg(not(target_os = "solana"))]
-    pub fn set_data(&mut self, data: &[u8]) -> Result<(), InstructionError> {
+    pub fn set_data(&mut self, data: Vec<u8>) -> Result<(), InstructionError> {
         self.can_data_be_resized(data.len())?;
         self.can_data_be_changed()?;
         self.touch()?;
-        if data.len() == self.account.data().len() {
-            self.account.data_as_mut_slice().copy_from_slice(data);
-        } else {
+
+        if data.len() != self.account.data().len() {
+            let mut accounts_resize_delta = self
+                .transaction_context
+                .accounts_resize_delta
+                .try_borrow_mut()
+                .map_err(|_| InstructionError::GenericError)?;
+            *accounts_resize_delta = accounts_resize_delta
+                .saturating_add((data.len() as i64).saturating_sub(self.get_data().len() as i64));
+        }
+        self.account.set_data(data);
+
+        Ok(())
+    }
+
+    /// Overwrites the account data and size (transaction wide).
+    ///
+    /// Call this when you have a slice of data you do not own and want to
+    /// replace the account data with it.
+    ///
+    /// If you have an owned buffer (eg Vec<u8>), use set_data().
+    #[cfg(not(target_os = "solana"))]
+    pub fn set_data_from_slice(&mut self, data: &[u8]) -> Result<(), InstructionError> {
+        self.can_data_be_resized(data.len())?;
+        self.can_data_be_changed()?;
+        self.touch()?;
+
+        if data.len() != self.account.data().len() {
             let mut accounts_resize_delta = self
                 .transaction_context
                 .accounts_resize_delta
                 .try_borrow_mut()
                 .map_err(|_| InstructionError::GenericError)?;
             *accounts_resize_delta = accounts_resize_delta
                 .saturating_add((data.len() as i64).saturating_sub(self.get_data().len() as i64));
-            self.account.set_data_from_slice(data);
         }
+        self.account.set_data_from_slice(data);
```

Comment:
I don't think this helps because `set_data_from_slice()` also implicitly calls `to_vec()` internally. It is thus the same as `set_data()`:
https://github.com/solana-labs/solana/blob/a3395a786aea208f81516c0e43d71c239bc0a631/sdk/src/account.rs#L546

IMO only an explicit split into `set_data_len()` and `get_data_mut().copy_from_slice()` can make a difference.

Reply:
I have a PR up for that as well, see https://github.com/solana-labs/solana/pull/27837